src/persist/src/azure.rs

```diff
@@ -0,0 +1,364 @@
+                    panic!("CI is supposed to run this test but something has gone wrong!");
+                }
+                return Ok(None);
+            }
+        };
+
+        let prefix = Uuid::new_v4().to_string();
+        let metrics = S3BlobMetrics::new(&MetricsRegistry::new());
+
+        let config = AzureBlobConfig::new(
+            EMULATOR_ACCOUNT.to_string(),
+            container_name.clone(),
+            prefix,
+            metrics,
+            Url::parse(&format!("http://localhost:40111/{}", container_name)).expect("valid url"),
+            Arc::new(ConfigSet::default()),
+        )?;
+
+        Ok(Some(config))
+    }
+
+    /// Returns a clone of Self with a new v4 uuid prefix.
+    pub fn clone_with_new_uuid_prefix(&self) -> Self {
+        let mut ret = self.clone();
+        ret.prefix = Uuid::new_v4().to_string();
+        ret
+    }
+}
+
+/// Implementation of [Blob] backed by Azure Blob Storage.
+#[derive(Debug)]
+pub struct AzureBlob {
+    metrics: S3BlobMetrics,
+    client: ContainerClient,
+    prefix: String,
+    _cfg: Arc<ConfigSet>,
+}
+
+impl AzureBlob {
+    /// Opens the given location for non-exclusive read-write access.
+    pub async fn open(config: AzureBlobConfig) -> Result<Self, ExternalError> {
+        if config.client.service_client().account() == EMULATOR_ACCOUNT {
+            // TODO: we could move this logic into the test harness.
+            // it's currently here because it's surprisingly annoying to
+            // create the container out-of-band
+            if let Err(e) = config.client.create().await {
+                warn!("Failed to create container: {e}");
+            }
+        }
+
+        let ret = AzureBlob {
+            metrics: config.metrics,
+            client: config.client,
+            prefix: config.prefix,
+            _cfg: config.cfg,
+        };
+
+        Ok(ret)
+    }
+
+    fn get_path(&self, key: &str) -> String {
+        format!("{}/{}", self.prefix, key)
+    }
+}
+
+#[async_trait]
+impl Blob for AzureBlob {
+    async fn get(&self, key: &str) -> Result<Option<SegmentedBytes>, ExternalError> {
+        let path = self.get_path(key);
+        let blob = self.client.blob_client(path);
+        let mut segments: Vec<MaybeLgBytes> = vec![];
+
+        let mut stream = blob.get().into_stream();
+        while let Some(value) = stream.next().await {
+            let response = match value {
+                Ok(v) => v,
+                Err(e) => {
+                    if let Some(e) = e.as_http_error() {
+                        if e.status() == StatusCode::NotFound {
+                            return Ok(None);
+                        }
+                    }
+
+                    return Err(ExternalError::from(anyhow!(
+                        "Azure blob get error: {:?}",
+                        e
+                    )));
+                }
+            };
+
+            let content_length = response.blob.properties.content_length;
```

Comment:
I'm a little scared of this value being invalid, e.g. 0, specifically the `buffer.extend_from_slice(...)` call will panic if there isn't enough room instead of resizing. Granted we do always [allocate a region of some minimum size](https://github.com/MaterializeInc/materialize/blob/main/src/ore/src/lgbytes.rs#L286), so the case of the `content_length` being 0 isn't to concerning. You could follow similar logic from the [S3 implementation](https://github.com/MaterializeInc/materialize/blob/main/src/persist/src/s3.rs#L523-L527), granted that's a lot more complex.

Sorry this is me thinking out loud a bit, overall we're good to go as-is, really the concern comes from my lack of knowledge about the reliability of Azure blob store!

Reply:
hm... not totally sure how to map the S3 implementation to this one. I'll ping you and follow up on this one

Reply:
updated with @ParkMyCar's patch